custom/enikshay/two_b_release_1/management/commands/enikshay_2b_case_properties.py

```diff
@@ -0,0 +1,496 @@
+        """Get all direct ancestors found in self.locations"""
+        ancestors_by_type = {location.location_type.code: location}
+        loc = location
+        while loc.parent_id and loc.parent_id in self.location_ids_by_pk:
+            parent = self.locations[self.location_ids_by_pk[loc.parent_id]]
+            ancestors_by_type[parent.location_type.code] = parent
+            loc = parent
+        return ancestors_by_type
+
+    def migrate(self):
+        person_ids = self.get_relevant_person_case_ids()
+        persons = self.get_relevant_person_case_sets(person_ids)
+        for person in with_progress_bar(persons, len(person_ids)):
+            self.migrate_person_case_set(person)
+
+    def get_relevant_person_case_ids(self):
+        location_owners = self.dto.get_descendants(include_self=True).location_ids()
+        return self.accessor.get_open_case_ids_in_domain_by_type(CASE_TYPE_PERSON, location_owners)
+
+    def get_relevant_person_case_sets(self, person_ids):
+        """
+        Generator returning all relevant cases for the migration, grouped by person.
+
+        This is a pretty nasty method, but it was the only way I could figure
+        out how to group the queries together, rather than performing multiple
+        queries per person case.
+        """
+        for person_chunk in chunked(person_ids, 100):
+            person_chunk = list(filter(None, person_chunk))
+            all_persons = {}  # case_id: PersonCaseSet
+            for person in self.accessor.get_cases(person_chunk):
+                # enrolled_in_private is blank/not set AND case_version is blank/not set
+                # AND owner_id is within the location set being migrated
+                if (person.get_case_property(ENROLLED_IN_PRIVATE) != 'true'
+                        and not person.get_case_property(CASE_VERSION)):
+                    all_persons[person.case_id] = PersonCaseSet(
+                        person=person,
+                        occurrences=[],
+                        episodes=[],
+                        tests=[],
+                        referrals=[],
+                        trails=[],
+                        drtb_hiv=[],
+                    )
+
+            referrals_and_occurrences_to_person = {}
+            type_to_bucket = {CASE_TYPE_OCCURRENCE: 'occurrences',
+                              CASE_TYPE_REFERRAL: 'referrals'}
+            for case in self.accessor.get_reverse_indexed_cases(
+                    [person_id for person_id in all_persons]):
+                bucket = type_to_bucket.get(case.type, None)
+                if bucket:
+                    for index in case.indices:
+                        if index.referenced_id in all_persons:
+                            getattr(all_persons[index.referenced_id], bucket).append(case)
+                            referrals_and_occurrences_to_person[case.case_id] = index.referenced_id
+                            break
```

Comment:
In this block I'm trying to match up occurrence and referral cases with their corresponding person cases (in bulk, which is not super straightforward).  My assumption is that there is at most one person case associated with occurrences and referrals.